File: solutions/cpp/rail-fence-cipher/1/rail_fence_cipher.cpp
#include "rail_fence_cipher.h"

#include <algorithm>
#include <cmath>
#include <cstdlib>
#include <utility>
#include <vector>
#include <array>

namespace rail_fence_cipher {

std::vector<std::pair<int, int>> get_mapping(int num_rails, int length) {
	std::vector<std::pair<int, int>> res {};
	int dec = 0;
	for (auto n = 0; n < num_rails; n++) {
		float middle = 0.5*((float)num_rails - ((num_rails & 1) ^ 1));
		int dist_from_middle = std::abs(n - 0.5*((float)num_rails - 1));
		std::array<int, 2> offsets {
			2*((int)middle - dist_from_middle),
			2*(num_rails - 1 - (int)middle + dist_from_middle),
		};

		bool flag = true;
		for (
			int enc = n;
			enc < length;
			enc += offsets[(flag + (offsets[0] != 0 and n > middle)) % 2],
			flag = (offsets[0] == 0) or not flag
		)
			res.push_back({dec++, enc});
	}

	return res;
}

std::string encode(const std::string& plaintext, int num_rails) {
	auto mapping = get_mapping(num_rails, plaintext.size());
	std::string encoded {};
	for (auto [dec, enc] : mapping)
		encoded += plaintext[enc];
	return encoded;
}

std::string decode(const std::string& ciphertext, int num_rails) {
	auto mapping = get_mapping(num_rails, ciphertext.size());
	std::sort(mapping.begin(), mapping.end(), [](auto pa, auto pb){ return pa.second < pb.second; });
	std::string decoded {};
	for (auto [dec, enc] : mapping)
		decoded += ciphertext[dec];
	return decoded;
}

}  // namespace rail_fence_cipher

File: solutions/lua/rail-fence-cipher/1/rail-fence-cipher.lua

local function list_rep (elt, n)
	local xs = {}
	for _ = 1, n do
		table.insert(xs, elt)
	end
	return xs
end

---@param num_rails integer
---@param text string
---@param is_encoding boolean
---@return string
local function code (num_rails, text, is_encoding)
	local coded = list_rep('A', #text)
	local dec = 0
	for n = 0, num_rails-1 do
		local middle = 0.5*(num_rails - ((num_rails & 1) ~ 1))
		local dist_from_middle = math.floor(math.abs(n - 0.5*(num_rails - 1)))
		---@type integer[]
		local offsets = {
			2*(math.floor(middle) - dist_from_middle),
			2*(num_rails - 1 - math.floor(middle) + dist_from_middle),
		}
		local flag = true
		local enc = n
		while enc < #text do
			if is_encoding then
				coded[dec+1] = text:sub(enc+1, enc+1)
			else
				coded[enc+1] = text:sub(dec+1, dec+1)
			end
			dec = dec + 1
			enc = enc + offsets[(flag ~= (offsets[1] ~= 0 and n > middle)) and 2 or 1]
			flag = (offsets[1] == 0) or not flag
		end
	end

	return table.concat(coded)
end

local function encode(plaintext, rail_count)
	return code(rail_count, plaintext, true)
end

local function decode(ciphertext, rail_count)
	return code(rail_count, ciphertext, false)
end

return { encode = encode, decode = decode }
